### PR TITLE
Add callsuper macro.

### LINKED ISSUE #18252: add @callsuper macro
As per https://github.com/JuliaLang/julia/pull/13123, we could add a `@supercall` macro that sweetens the interface to `invoke` and makes it more intuitive. My current implementation:

``` jl
macro callsuper(ex)
    ex.head == :call || error("@invoke requires a call expression")
    args = ex.args[2:end]
    types = Symbol[]
    vals = Symbol[]
    blk = quote end
    for arg in args
            val = gensym()
            typ = gensym()
            push!(vals, val)
            push!(types, typ)
            if isa(arg,Expr) && arg.head == :(::) && length(arg.args) == 2
                push!(blk.args, :($typ = $(esc(arg.args[2]))))
                push!(blk.args, :($val = $(esc(arg.args[1]))::$typ))
            else
                push!(blk.args, :($val = $(esc(arg))))
                push!(blk.args, :($typ = typeof($val)))
            end
    end
    push!(blk.args, :(invoke($(esc(ex.args[1])), ($(types...),), $(vals...))))
    return blk
end
```

Support for keyword arguments could be taken from https://github.com/JuliaLang/julia/pull/11165.

### PR DESCRIPTION
**Pull request by simonbyrne**
Fixes #18252 and #7045.

### CODE REVIEW

**Comment by simonbyrne** [comment]
Feel free to bikeshed names.

File: base/util.jl [hunk 1]
```diff
@@ -750,3 +750,47 @@ kwdef_val(::Type{Cwstring}) = Cwstring(C_NULL)
...
+macro callsuper(ex)
+    ex.head == :call || error("@invoke requires a call expression")
+    fname = ex.args[1]
+    args = ex.args[2:end]
+    types = Symbol[]
+    vals = Symbol[]
+    kwargs = Any[]
+    blk = quote end
+    for arg in args
+        if isa(arg,Expr) && arg.head == :kw
```
Comment (yuyichao) [inline base/util.jl]: This is not entirely correct. A few cases that should be handled and possibly tested are. (These are the reason I started but didn't finish the pr yet...)

1. Evaluation order

    Currently `:parameters` are always evaluated the first. It's not necessarily the best choice but this should be consisent with normal calls.

2. `:parameters` can have `...`

3. `:parameters` can have non `:kw` in it

4. positional arguments can also have `...`

5. if `parameter` only have `...`s and all of them are empty, the non-kw version should be called.

**Review by yuyichao — CHANGES_REQUESTED** [review]

File: base/util.jl (hunk 1 above)
Comment (simonbyrne) [inline base/util.jl]: what non-`:kw` can a `:parameters` have other than `...`?

Comment (yuyichao) [inline base/util.jl]: Any, iterable objects

```jl
julia> f(;kws...) = kws
f (generic function with 1 method)

julia> j = [:j, :k]
2-element Array{Symbol,1}:
 :j
 :k

julia> f(;:a=>:b, (:c, :d), [:e, :f], (:g, :h, :i), j)
5-element Array{Any,1}:
 (:a,:b)
 (:c,:d)
 (:e,:f)
 (:g,:h)
 (:j,:k)
```

File: base/util.jl [hunk 2]
```diff
@@ -750,3 +750,47 @@ kwdef_val(::Type{Cwstring}) = Cwstring(C_NULL)
...
+                push!(kwargs, QuoteNode(aarg.args[1]))
+                push!(kwargs, esc(aarg.args[2]))
+            end
+        else
+            val = gensym()
+            typ = gensym()
+            push!(vals, val)
+            push!(types, typ)
+            if isa(arg,Expr) && arg.head == :(::) && length(arg.args) == 2
+                push!(blk.args, :($typ = $(esc(arg.args[2]))))
```
Comment (yuyichao) [inline base/util.jl]: Also, `val` should be evaluated before the type.

File: base/util.jl [hunk 3]
```diff
@@ -750,3 +750,47 @@ kwdef_val(::Type{Cwstring}) = Cwstring(C_NULL)
...
+                push!(blk.args, :($val = $(esc(arg.args[1]))::$typ))
+            else
+                push!(blk.args, :($val = $(esc(arg))))
+                push!(blk.args, :($typ = typeof($val)))
+            end
+        end
+    end
+    if isempty(kwargs)
+        push!(blk.args, :(invoke($(esc(fname)), Tuple{$(types...)}, $(vals...))))
```
Comment (yuyichao) [inline base/util.jl]: And `fname` should be evaluated before the arguments.

Comment (simonbyrne) [inline base/util.jl]: Actually, `fname` evaluation is a bit weird:
```
julia> f(args...; kwargs...) = (args,kwargs)
f (generic function with 1 method)

julia> (println("f"); f)((println("a"); "a"); [:y,(println("b"); "b")])
b
f
f
a
(("a",),Any[(:y,"b")])

julia> (println("f"); f)((println("a"); "a"); [:y => (println("b"); "b")]...)
b
f
a
f
f
(("a",),Any[(:y,"b")])
```

File: base/util.jl (hunk 1 above)
Comment (StefanKarpinski) [inline base/util.jl]: @yuyichao, wouldn't be easier all around if you just did this PR? There was a similar back and forth when I proposed a possible implementation

**Comment by StefanKarpinski** [comment]
I like the name, @simonbyrne. Thanks for pushing this forward – it's a much nicer interface than bare `invoke`.

File: base/util.jl (hunk 1 above)
Comment (yuyichao) [inline base/util.jl]: Maybe. I started a implementation but haven't finished it yet (collecting data of what input should/can be supported and what generic fallback path I can use).....

File: base/util.jl (hunk 3 above)
Comment (yuyichao) [inline base/util.jl]: This is a lowering bug. The lowering code should emit a temporary variable to hold `f` if it cannot proof that evaluating `f` is "pure".

**Comment by JeffBezanson** [comment]
IIUC, this doesn't fully fix #7045 since it doesn't fix the case of using `invoke` directly.

**Comment by jlumpe** [comment]
This would be a really nice feature, is anyone still working on it?

**Comment by Keno** [comment]
If I understand correctly this is superseded by #38438. Let me know if I misunderstood.

**Comment by simonbyrne** [comment]
Indeed it is.

**Pull request closed without merging**